docs/adr/0006-consensus-governance.md

```diff
@@ -0,0 +1,478 @@
+# ADR 0006: Consensus Governance
+
+## Changelog
+
+- 2020-10-27: Voting period in epochs, min upgrade cancellation difference,
+  failed proposal state
+- 2020-10-16: Initial draft
+
+## Status
+
+Proposed
+
+## Context
+
+Currently the consensus layer does not contain any on-chain governance
+mechanism so any network upgrades need to be carefully coordinated off-chain.
+An on-chain governance mechanism would allow upgrades to be handled in a more
+controlled (and automatable) manner without introducing the risk of corrupting
+state.
+
+## Decision
+
+This proposal introduces a minimal on-chain governance mechanism where anyone
+can submit governance proposals and the validators can vote where one base unit
+of delegated stake counts as one vote.
+
+The high-level overview is as follows:
+
+- **A new governance API** is added to the consensus layer and its Tendermint
+  based implementation. It supports transactions for submitting proposals and
+  voting on proposals. It supports queries for listing current proposals and
+  votes for any given proposal.
+
+- **Two governance proposal kinds are supported**, a consensus layer upgrade
+  proposal (where the content is basically the existing upgrade descriptor) and
+  the cancellation of a pending upgrade.
+
+A proposal is created through a _submit proposal_ transaction and requires a
+minimum deposit (which is later refunded in case the proposal passes). Once a
+proposal is successfully submitted the voting period starts. Entities that are
+part of the validator set may cast votes for the proposal. After the voting
+period completes, the votes are tallied and the proposal either passes or is
+rejected.
+
+In case the proposal passes, the actions specified in the content of the propsal
+are executed. Currently the only actions are scheduling of an upgrade by
+publishing an upgrade descriptor or cancelling a previously passed upgrade.
+
+### State
+
+#### Staking
+
+This proposal adds the following consensus layer state in the staking module:
+
+- **Governance deposits account balance (`0x59`)**, similar to the common pool.
+
+#### Governance
+
+This proposal adds the following consensus layer state in the governance module:
+
+- **Next proposal identifier (`0x80`)**
```

Comment:
If we follow the existing convention, I think these identifiers will start at `0x90` (and the governance app id will be `0x09`) since `0x08` is in use by the new [improved beacon already](https://github.com/oasisprotocol/oasis-core/pull/3180/files).

Reply:
Hm, so I looked at the beacon when deciding on this and it uses the prefix from the previous beacon/epochtime apps which seems like the right thing to do:
https://github.com/oasisprotocol/oasis-core/blob/ae0e6719ffa2cc8e67a83c25943321a89e957cc7/go/consensus/tendermint/apps/beacon/state/state.go#L16-L32

The app id for the beacon should be changed to `0x04` to be consistent.

Reply:
Ok yeah I also saw the comment in the beacon PR :+1: